plugins/dell-dock2/fu-dell-dock2-rtshub.c

```diff
@@ -0,0 +1,534 @@
+{
+	FuHIDCmdBuffer cmd_buffer = {
+	    .cmd = RTSHUB_CMD_WRITE_DATA,
+	    .ext = RTSHUB_EXT_RESET_TO_FLASH,
+	    .dwregaddr = 0,
+	    .bufferlen = 0,
+	};
+	guint8 buf[DELL_DOCK2_RTSHUB_BUFFER_SIZE] = {0};
+
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    0x0, /* dst */
+			    (const guint8 *)&cmd_buffer,
+			    sizeof(cmd_buffer),
+			    0x0, /* src */
+			    sizeof(cmd_buffer),
+			    error))
+		return FALSE;
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      0x0,
+				      buf,
+				      sizeof(buf),
+				      DELL_DOCK2_RTSHUB_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to soft reset: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_rtshub_write_flash(FuDellDock2RtsHub *self,
+				 guint32 addr,
+				 const guint8 *data,
+				 guint16 data_sz,
+				 GError **error)
+{
+	FuHIDCmdBuffer cmd_buffer = {
+	    .cmd = RTSHUB_CMD_WRITE_DATA,
+	    .ext = RTSHUB_EXT_WRITEFLASH,
+	    .dwregaddr = GUINT32_TO_LE(addr),
+	    .bufferlen = GUINT16_TO_LE(data_sz),
+	};
+	guint8 buf[DELL_DOCK2_RTSHUB_BUFFER_SIZE] = {0};
+
+	g_return_val_if_fail(data_sz <= 128, FALSE);
+	g_return_val_if_fail(data != NULL, FALSE);
+	g_return_val_if_fail(data_sz != 0, FALSE);
+
+	/* vendor command */
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    0x0, /* dst */
+			    (const guint8 *)&cmd_buffer,
+			    sizeof(cmd_buffer),
+			    0x0, /* src */
+			    sizeof(cmd_buffer),
+			    error))
+		return FALSE;
+
+	/* data payload */
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    DELL_DOCK2_RTSHUB_WRITE_FLASH_OFFSET_DATA, /* dst */
+			    data,
+			    data_sz,
+			    0x0, /* src */
+			    data_sz,
+			    error))
+		return FALSE;
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      0x0,
+				      buf,
+				      sizeof(buf),
+				      DELL_DOCK2_RTSHUB_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to write flash @%08x: ", (guint)addr);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_rtshub_write_fw(FuDevice *device,
+			      FuFirmware *firmware,
+			      FuProgress *progress,
+			      FwupdInstallFlags flags,
+			      GError **error)
+{
+	FuDellDock2RtsHub *self = FU_DELL_DOCK2_RTSHUB(device);
+	g_autoptr(GInputStream) stream = NULL;
+	g_autoptr(FuChunkArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 46, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 52, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, "reset");
+
+	/* get default image */
+	stream = fu_firmware_get_stream(firmware, error);
+	if (stream == NULL)
+		return FALSE;
+
+	/* unlock the target */
+	if (!fu_dell_dock2_request_modify_lock(device, self->unlock_target, TRUE, error))
```

Comment:
There is an error at the time it is trying to unlock the device through ec, typically says the device has been disconnected.
```
failed to write-firmware: Failed to unlock device 8: write over HID-I2C failed: device recovery not possible: failed to SetReport: USB error: No such device (it may have been disconnected) [-4]
```

I thought the method here opens the device (or reuse the handler if it has been opened), any idea why it complains no such device?

[x.log](https://github.com/user-attachments/files/16323024/x.log)

Reply:
I'm seeing a bunch of libusb errors like this:
`libusb: error [udev_hotplug_event] ignoring udev action change`

Maybe fwupd is holding the events?

Reply:
Is `FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG` flag expecting a device reset that some devices will be removed and reconnected back?

https://github.com/fwupd/fwupd/pull/7499/files#diff-a47c06516017112603cf2e7d8b0222093f6d880ed29b7dfabb0745a0687f5920R334

Reply:
> Is FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG flag expecting a device reset that

if you set that flag (on as many devices as you want) then fwupd will wait for them all to disappear and reappear before doing the next step.